src/NumericalAlgorithms/LinearOperators/PowerMonitors.cpp

```diff
@@ -54,6 +55,94 @@ std::array<DataVector, Dim> power_monitors(
   return result;
 }
 
+double relative_truncation_error(const DataVector& input_power_monitors,
+                                 const size_t num_modes) {
+  ASSERT(num_modes <= input_power_monitors.size(),
+         "Number of modes needs less or equal than the number of input power "
+         "monitors");
+  ASSERT(2_st <= num_modes,
+         "Number of modes needs to be larger or equal than 2.");
+  // Compute weighted average and total sum in the current dimension
+  double weighted_average = 0.0;
+  double weight_sum = 0.0;
+  double weight_value = 0.0;
+  for (size_t i = 0; i < num_modes; ++i) {
+    // Compute current weight
+    weight_value = exp(-square(i - static_cast<double>(num_modes) + 0.5));
+    // Add weighted power monitor
+    const auto mode = gsl::at(input_power_monitors, i);
+    if (not equal_within_roundoff(mode, 0.0)) {
+      weighted_average += weight_value * log10(mode);
+    }
+    else {
+      // If zero within roundoff set a maximum resolution
+      weighted_average += - 34.0 * weight_value;
```

Comment:
Why 34? Doesn't this term completely dominate the result if a mode is zero to numerical roundoff? Maybe we should ignore zero modes in the average or something like that. Do you have an idea?

Reply:
I can imagine the case when, for instance, even modes are zero (corresponding to -34) by some symmetry. Then this term might be large, but should average with the non zero ones giving some value in between --since at the end we divide by the total sum of the weights. 

But maybe instead of -34 I could use log10 of `std::numeric_limits<double>::epsilon()` ?

Reply:
I've also been playing with the following variation which drops modes from the sum if they are zero within roundoff (and adjusting the weights when we drop modes on the higher end). 

I've noticed one issue that implementation. (I put it here since I'm not yet convinced by it.) 

For the case of a test function that is equal to a constant (and for which modes of order >1 are zero) I cannot apply the Szilagyi formula anymore. Returning zero doesn't seem right since for this function one should easily get good resolution.
I am also worried that (for similar reasons) functions that are well resolved by few modes will seem to have large relative truncation errors that don't improve with the number of modes.

```
double relative_truncation_error(const DataVector& input_power_monitors,
                                 const size_t num_modes) {
  ASSERT(num_modes <= input_power_monitors.size(),
         "Number of modes needs less or equal than the number of input power "
         "monitors");
  ASSERT(2_st <= num_modes,
         "Number of modes needs to be larger or equal than 2.");

  // Maximum between the first two power monitors
  double leading_term = std::max(gsl::at(input_power_monitors, 0_st),
                                 gsl::at(input_power_monitors, 1_st));
  if (not equal_within_roundoff(leading_term, 0.0)) {
    leading_term = log10(leading_term);
  } else {
    // If leading term is zero within roundoff
    // cannot compute relative error using Eq. (57) of Ref.
    // \cite Szilagyi2014fna
    return 0.0;
  }

  // Compute weighted average and total sum in the current dimension
  double weighted_average = 0.0;
  double weight_sum = 0.0;
  double weight_value = 0.0;
  bool highest_non_zero_mode = false;

  int num_modes_in_average = num_modes;
  for (int i = num_modes_in_average - 1; 0 <= i; --i) {
    // Get power monitor
    const double mode = gsl::at(input_power_monitors, i);
    if (not equal_within_roundoff(mode, 0.0)) {
      // When we find the highest non zero mode this flag is set to true
      // and the number of modes considered in the average is fixed
      highest_non_zero_mode = true;
      // Compute current weight
      weight_value = exp(-square(i - num_modes_in_average + 0.5));
      weighted_average += weight_value * log10(mode);
    } else if (highest_non_zero_mode and num_modes_in_average > 2) {
      // Skip intermediate modes within roundoff error from zero
      continue;
    } else if (not highest_non_zero_mode and num_modes_in_average > 2) {
      // If the last mode is within roundoff error from zero remove from the
      // modes considered for the average
      // The new total number of modes is now
      --num_modes_in_average;
      // Reset values and restart
      i = num_modes_in_average - 1;
      weighted_average = 0.0;
      weight_sum = 0.0;
      continue;
    } else {
      // If all modes higher than the leading modes are within roundoff
      // no relative error can be computed
      return 0.0;
    }
    // Add term to weighted sum
    weight_sum += weight_value;
  }
  weighted_average /= weight_sum;

  // Compute relative truncation error
  return leading_term - weighted_average;
}
```

Reply:
Please drop the floor. The floor adds an absolute error tolerance to this function and has in practice not been needed for several thousand BBH simulations. I think we should first do what is working in SpEC, then we can experiment with different methods of estimating the truncation error later.

I'm not so worried about the highest modes being zero because I thought that's what the `num_modes` argument is used to fix. I think that's these "pile-up modes" that you're talking about, and `num_modes` passed into the truncation error estimate should ignore the pile-up modes. If you're worried about a mode being identically zero, I think adding an `ASSERT` for that is reasonable :)

I think symmetry cases like zero even modes are also going to be rare in practice. I can't think of any astrophysical system that has such a high degree of symmetry in polynomials.